cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -328,12 +328,12 @@ class DatasetWriterDirectoryQueue : public util::AsyncDestroyable {
   uint64_t rows_written() const { return rows_written_; }
 
   void PrepareDirectory() {
-    init_future_ =
-        DeferNotOk(write_options_.filesystem->io_context().executor()->Submit([this] {
+    init_future_ = DeferNotOk(
+        write_options_.filesystem->io_context().executor()->Submit([this]() -> Future<> {
```

Comment:
It would probably be a good addition to allow executor submission tasks to return futures but I don't think this actually works today.  The future you are returning from the task will be discarded without being awaited.

I think you'll need to change it to...

`init_future_ = SubmitTask(CreateDir).Then(DeleteDirContentsAsync);`